sample-programs/erc20/erc20.scm

```diff
@@ -0,0 +1,129 @@
+    (if (eq? balance 'null) 0 balance)))
+
+;; Gets the amount of tokens that an owner allowed to a spender.
+(define (allowance-of (spender principal) (owner principal))
+  (let ((allowance (get allowance 
+      (fetch-entry allowances (tuple 
+        (owner owner) 
+        (spender spender))))))
+    (if (eq? allowance 'null) 0 allowance)))
+
+;; Credits balance of a specified principal.
+(define (credit-balance! (account principal) (amount int))
+  (if (<= amount 0)
+    'false
+    (let ((current-balance (balance-of account)))
+      (begin
+        (set-entry! balances 
+          (tuple (owner account))
+          (tuple (balance (+ amount current-balance)))) 
+        'true)))) ;; Overflow management?
+
+;; Debits balance of a specified principal.
+(define (debit-balance! (account principal) (amount int))
+  (let ((balance (balance-of account)))
+    (if (or (> amount balance) (<= amount 0))
+      'false
+      (begin
+        (set-entry! balances
+          (tuple (owner account))
+          (tuple (balance (- balance amount))))
+        'true))))
+
+;; Transfers tokens to a specified principal.
+(define (transfer! (sender principal) (recipient principal) (amount int))
+  (if (and (credit-balance! recipient amount) (debit-balance! sender amount)) 
+    'true
+    'false))
+
+;; Decrease allowance of a specified spender.
+(define (decrease-allowance! (spender principal) (owner principal) (amount int))
+  (let ((allowance (allowance-of spender owner)))
+    (if (or (> amount allowance) (<= amount 0))
+      'false
+      (begin
+        (set-entry! allowances
+          (tuple (spender spender) (owner owner))
+          (tuple (allowance (- allowance amount))))
+        'true))))
+
+;; Internal - Increase allowance of a specified spender.
+(define (increase-allowance! (spender principal) (owner principal) (amount int))
+  (let ((allowance (allowance-of spender owner)))
+    (if (<= amount 0)
+      'false
+      (begin
+        (set-entry! allowances 
+          (tuple (spender spender) (owner owner))
+          (tuple (allowance (+ allowance amount))))
+        'true)))) 
+
+;; Public functions
+
+;; Transfers tokens to a specified principal.
+(define-public (transfer (recipient principal) (amount int))
+  (transfer! tx-sender recipient amount))
+
+;; Transfers tokens to a specified principal, performed by a spender
+(define-public (transfer-from (owner principal) (recipient principal) (amount int))
+  (let ((allowance (allowance-of tx-sender owner)))
+    (if (or (> amount allowance) (<= amount 0))
+      'false
+      (begin
+        (transfer! owner recipient amount)
+        (decrease-allowance! tx-sender owner amount)))))
+
+;; Update the allowance for a given spender
+(define-public (approve (spender principal) (amount int))
+  (if (<= amount 0)
+    'false
+    (begin
+      (increase-allowance! spender tx-sender amount))))
+
+;; Mint new tokens.
+(define-public (mint! (account principal) (amount int))

```

Comment:
There aren't any checks on `account`.  Is anyone allowed to mint any number of tokens?

Reply:
You're right, I did cut a corner and should have write an explanation.
`mint!` should be a private method, and we should use mint! in a `constructor` / `init` method (something we don't have at this point - it's on my list of feedback / questions).

If having a native `constructor` / `init` is not on the roadmap, then I'll write an `on-chain` method with the right checks (should only be called by the contract's publisher, should run once, and all the other methods can't be called before this one).

Reply:
There's no need for constructors or init methods in our smart contracts.  You can just add code at the very end that will be executed when the contract is instantiated (since to the VM, it's all the same).

Example:

```
 (define-map pythags ((a int) (b int) (c int)) ((result bool)))

(define-public (test-pythag-triple (a int) (b int) (c int))
   (let
     (
         (a2 (* a a))
         (b2 (* b b))
         (c2 (* c c))
     )
     (begin
       (if (eq? (+ a2 b2) c2)
           (set-entry! pythags (tuple (a a) (b b) (c c)) (tuple (result 'true)))
           (set-entry! pythags (tuple (a a) (b b) (c c)) (tuple (result 'false)))
       )
       'true
     )
   )
)

(define-public (is-pythag-triple (a int) (b int) (c int))
    (let 
      (
        (result (get result (fetch-entry pythags (tuple (a a) (b b) (c c)))))
      )
      (begin
        (if (eq? result 'null)
           'false
           result
        )
      )
   )
)

(begin
  (test-pythag-triple 3 4 5)
  (test-pythag-triple 1 2 3)
  'null
)
```

Here, `test-pythag-triple` will be called twice with `3 4 5` and `1 2 3` when the contract is instantiated.

Reply:
Sounds good, thanks!

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/955/commits/64b2652214b20518a539c910648e2a3d02f0ee93